### PR TITLE
Add option to use Auryn as a container

### PR DESCRIPTION
**Pull request by shadowhand**
Auryn provides a powerful reflection-based injector. Although Auryn
itself does not implement PSR-11 a wrapper does exist for it.

Depends on changes from #160.

### CODE REVIEW

**Comment by shadowhand** [comment]
I played around with `composer create-package` but couldn't figure out how to make it target my PR. A couple of other things to note:

- the screenshot in the README will be out of date when this is merged
- the changelog version was a guess

File: src/ExpressiveInstaller/Resources/config/container-auryn.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
+<?php
+
+use Auryn\Injector;
```
Comment (Ocramius) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: `Auryn` is explicitly used in the file, so it should also be explicity defined as dependency, IMO

File: src/ExpressiveInstaller/Resources/config/container-auryn.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// Load configuration
+$config = require __DIR__ . '/config.php';
+
+// Create injector
+$injector = new Injector();
+
+// Configuration will be used globally
+$injector->share('config')->delegate('config', function () use ($config) {
+    // Auryn requires that all injections resolve to an object.
+    // Wrap the config array with an ArrayObject to appease the gods.
```
Comment (Ocramius) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: This is something that may interest @xtreamwayz: I thought we nuked `Aura.DI` at some point because of this?

File: src/ExpressiveInstaller/Resources/config/container-auryn.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
+<?php
+
+use Auryn\Injector;
+use Northwoods\Container\InjectorContainer;
+use Psr\Container\ContainerInterface;
+
+// Load configuration
+$config = require __DIR__ . '/config.php';
```
Comment (Ocramius) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: Consider wrapping the entire body of this file in an IIFE:

```php
(function () {
}());
```

File: src/ExpressiveInstaller/Resources/config/container-auryn.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+$injector->share(ContainerInterface::class);
+
+// Container will resolve by injector
+$injector->alias(ContainerInterface::class, InjectorContainer::class);
+
+// Container will be created from injector
+$injector->define(InjectorContainer::class, [':injector' => $injector]);
+
+// Inject services
+if (! empty($config['dependencies']['services'])) {
```
Comment (Ocramius) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: A small utility in `Northwoods\Container` (something like `(new LoadServiceManagerAlikeConfiguration)($container, $config)` could reduce this code by a lot, and would allow testing the produced injector properly

File: src/ExpressiveInstaller/config.php [hunk 5]
```diff
@@ -39,7 +40,7 @@
     'questions' => [
         'container' => [
             'question'               => 'Which container do you want to use for dependency injection?',
-            'default'                => 3,
+            'default'                => 4,
```
Comment (Ocramius) [inline src/ExpressiveInstaller/config.php]: Instead of changing the option, move `northwoods/container` to the last key. This also avoids breaking existing automation by consumers of this package.

File: test/ExpressiveInstallerTest/ContainersTest.php [hunk 6]
```diff
@@ -98,12 +99,15 @@ public function containerProvider()
             'aura-minimal'    => [OptionalPackages::INSTALL_MINIMAL, 1, 2, 'minimal-files', 404, AuraContainer::class],
             'aura-flat'       => [OptionalPackages::INSTALL_FLAT,    1, 2, 'copy-files', 200, AuraContainer::class],
             'aura-modular'    => [OptionalPackages::INSTALL_MODULAR, 1, 2, 'copy-files', 200, AuraContainer::class],
-            'pimple-minimal'  => [OptionalPackages::INSTALL_MINIMAL, 2, 2, 'minimal-files', 404, PimpleContainer::class],
-            'pimple-flat'     => [OptionalPackages::INSTALL_FLAT,    2, 2, 'copy-files', 200, PimpleContainer::class],
-            'pimple-modular'  => [OptionalPackages::INSTALL_MODULAR, 2, 2, 'copy-files', 200, PimpleContainer::class],
-            'zend-sm-minimal' => [OptionalPackages::INSTALL_MINIMAL, 3, 2, 'minimal-files', 404, ZendServiceManagerContainer::class],
```
Comment (Ocramius) [inline test/ExpressiveInstallerTest/ContainersTest.php]: Same as above: let's reduce diffs by reducing the amount of changed keys. Adding-only is preferrable

File: test/AppTest/Action/HomePageFactoryTest.php [hunk 7]
```diff
@@ -4,8 +4,8 @@
 
 use App\Action\HomePageAction;
 use App\Action\HomePageFactory;
-use Interop\Container\ContainerInterface;
 use PHPUnit\Framework\TestCase;
+use Psr\Container\ContainerInterface;
```
Comment (Ocramius) [inline test/AppTest/Action/HomePageFactoryTest.php]: What's the reason for this change?

File: test/ExpressiveInstallerTest/ProjectSandboxTrait.php [hunk 8]
```diff
@@ -11,8 +11,8 @@
 use App\Action\PingAction;
 use DirectoryIterator;
 use ExpressiveInstaller\OptionalPackages;
-use Interop\Container\ContainerInterface;
 use PHPUnit\Framework\Assert;
+use Psr\Container\ContainerInterface;
```
Comment (Ocramius) [inline test/ExpressiveInstallerTest/ProjectSandboxTrait.php]: What's the reason for this change?

File: src/ExpressiveInstaller/Resources/config/container-auryn.php (hunk 2 above)
Comment (shadowhand) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: [aura.da still does it](https://github.com/zendframework/zend-expressive-skeleton/blob/d6d44f34dfc9a1eef1cecc8c99538ff2df982662/src/ExpressiveInstaller/Resources/src/ExpressiveAuraConfig.php#L54)

File: src/ExpressiveInstaller/Resources/config/container-auryn.php (hunk 4 above)
Comment (shadowhand) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: Can do.

File: test/AppTest/Action/HomePageFactoryTest.php (hunk 7 above)
Comment (shadowhand) [inline test/AppTest/Action/HomePageFactoryTest.php]: #160

File: test/ExpressiveInstallerTest/ProjectSandboxTrait.php (hunk 8 above)
Comment (shadowhand) [inline test/ExpressiveInstallerTest/ProjectSandboxTrait.php]: #160

File: src/ExpressiveInstaller/Resources/config/container-auryn.php (hunk 2 above)
Comment (xtreamwayz) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: @Ocramius Aura.DI still requires an object. I just moved it to the Aura.DI container file as @shadowhand linked.

File: src/ExpressiveInstaller/Resources/config/container-auryn.php (hunk 4 above)
Comment (shadowhand) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: northwoods/container#4

File: src/ExpressiveInstaller/config.php (hunk 5 above)
Comment (shadowhand) [inline src/ExpressiveInstaller/config.php]: All the rest of them are in alphabetical order... shouldn't the default one be first? 🎩

Comment (weierophinney) [inline src/ExpressiveInstaller/config.php]: Probably, but I'd not considered automation when we first did this, and now we need to. :smile:

We can re-order them for v3 of the skeleton, as that's a time when folks can expect automation might break.

File: src/ExpressiveInstaller/Resources/config/container-auryn.php [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+/** @var \Auryn\Injector */
+$injector = $builder->build();
+
+$injector->share('config')->delegate('config', function () use ($config) {
+    // Auryn requires that all injections resolve to an object.
+    // Wrap the config array with an ArrayObject to appease the gods.
+    return new ArrayObject($config);
+});
+
+return $injector->make(ContainerInterface::class);
```
Comment (shadowhand) [inline src/ExpressiveInstaller/Resources/config/container-auryn.php]: @Ocramius how's this look?

**Comment by guilhermeaiolfi** [comment]
I tried this pull request and I'm getting the same error as described in the comment above.

**Comment by shadowhand** [comment]
The bug you are both experiencing is due to #160.

**Comment by shadowhand** [comment]
@xtreamwayz @guilhermeaiolfi try now, rebase should have resolved the issue.

File: src/ExpressiveInstaller/config.php [hunk 10]
```diff
@@ -4,6 +4,7 @@
     'packages' => [
         'aura/di'                                        => '^3.2',
         'filp/whoops'                                    => '^2.1.7',
+        'northwoods/container'                           => '^1.0.2',
```
Comment (webimpress) [inline src/ExpressiveInstaller/config.php]: `^1.1` here

**Comment by guilhermeaiolfi** [comment]
How should I proceed is other packages use ```Interlop\Container\ContainerInterface``` as of ```PlatesRendererFactory```? I get the same error for it. I tried without success adding:

```php
$injector->alias(Interlop\Container\ContainerInterface::class, Psr\Container\ContainerInterface::class);
```

**Comment by guilhermeaiolfi** [comment]
I've found two inconsistencies that I had to workaround to use auryn.

First, when using ```UrlHelper``` it asks for ```RouterInterface``` in its constructor. I had to call ```$injector->share(RouterInterface::class)``` in container.php so it would work the same way as it was when using Pimple. Otherwise it would create a new instance of the Router.

Second, the Pimple ```has``` method works differently than auryn's. For auryn if the class exists but was not mapped anywhere it returns true, for Pimple it returns false. The side effect appears when using ```UrlHelper``` (maybe in other places too): pimple registers Plates extensions using ```PlatesEngineFactory```. For auryn because of this, it creates a ```PlatesEngine``` and ignores the factory and no extension is registered.

**Comment by xtreamwayz** [comment]
> How should I proceed is other packages use Interlop\Container\ContainerInterface as of PlatesRendererFactory? - @guilhermeaiolfi 
Could you try this in `composer.json`

```json
"conflict": {
    "container-interop/container-interop": "<1.2.0"
}
```
This should force  a PSR-11 compatible container-interop interface.

**Comment by guilhermeaiolfi** [comment]
I just tested your suggestion, @xtreamwayz. It didn't work.

**Comment by shadowhand** [comment]
@xtreamwayz @Ocramius I believe the issue you were seeing with shares has been resolved by [version 1.2.0](https://github.com/northwoods/container/blob/master/CHANGELOG.md#120).

**Comment by shadowhand** [comment]
@guilhermeaiolfi regarding this:

> For auryn if the class exists but was not mapped anywhere it returns true, for Pimple it returns false.

I don't see how that can be addressed. Auryn will happily create any class that is available, whereas Pimple cannot. The only option is to register a delegate to force the usage of the factory.

**Comment by xtreamwayz** [comment]
Merged into release-3.0.0. Thanx @shadowhand!
